llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -5921,6 +5921,61 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
 
     break;
   }
+  case Instruction::BitCast: {
+    const Value *Src;
+    if (!match(Op, m_ElementWiseBitCast(m_Value(Src))) ||
+        !Src->getType()->isIntOrIntVectorTy())
+      break;
+
+    const Type *Ty = Op->getType()->getScalarType();
+    KnownBits Bits(Ty->getScalarSizeInBits());
+    computeKnownBits(Src, DemandedElts, Bits, Depth + 1, Q);
+
+    // Transfer information from the sign bit.
+    if (Bits.isNonNegative())
+      Known.signBitMustBeZero();
+    else if (Bits.isNegative())
+      Known.signBitMustBeOne();
+
+    if (Ty->isIEEE()) {
+      // IEEE floats are NaN when all bits of the exponent plus at least one of
+      // the fraction bits are 1. This means:
+      //   - If we assume unknown bits are 0 and the value is NaN, it will
+      //     always be NaN
+      //   - If we assume unknown bits are 1 and the value is not NaN, it can
+      //     never be NaN
+      if (APFloat(Ty->getFltSemantics(), Bits.One).isNaN())
+        Known.KnownFPClasses = fcNan;
+      else if (!APFloat(Ty->getFltSemantics(), ~Bits.Zero).isNaN())
+        Known.knownNot(fcNan);
+
+      // Build KnownBits representing Inf and check if it must be equal or
+      // unequal to this value.
+      auto InfKB = KnownBits::makeConstant(
+          APFloat::getInf(Ty->getFltSemantics()).bitcastToAPInt());
+      InfKB.Zero.clearSignBit();
+      if (const auto InfResult = KnownBits::eq(Bits, InfKB)) {
```

Comment:
Think you need to unset signbit in `Bits` as well.

Reply:
Then ofc reset later, that or test against inf and -inf.

Reply:
Likewise below

Reply:
I don't think so. `KnownBits::eq` will return `false` if the inputs cannot be equal and `std::nullopt` if the may or may not be equal (in this case it cannot return `true` because `InfKB` is not fully known). Clearing the sign bit of `Bits` won't change the result either way.

Reply:
If `Bits` are -inf, won't you get known-false incorrectly? Or am I missing something?

Reply:
Oh yeah are clearing the signbit.

This is a bit odd, I don't think you can ever get `KnownBits::eq` to ever not be `std::nullopt` with unknown bits....

Likewise if the signbit is known in `Bits`, the follow up `==` will never be true...

Reply:
>  I don't think you can ever get KnownBits::eq to ever not be std::nullopt with unknown bits....

It can never return true, but it certainly return false if any bit is know to be 1 in one value and 0 in the other.

> Likewise if the signbit is known in Bits, the follow up == will never be true...

That `==` case is there to handle the case where the value is not a constant (will be folded) but it is still known that it must be inf, the only way this is possible is if the known bits are exactly `? 11111111 00000000000000000000000`.

There are examples of both of these cases in my tests.

Reply:
I see, I was missing it will constant fold if signbit is known.